File: benchmark/producer.py
"""Producer: enqueues single records at target rate, or batch-by-batch on signal (round-robin)."""
from __future__ import annotations

import json
import queue
import random
import time
from multiprocessing.synchronize import Synchronized
from typing import Any, Union

from .config import INSERTION_SENTINEL
from .patient_generator import generate_one_patient, DUPLICATE_RATIO

Record = tuple[str, str, str, bool]

# Message type for rows built from generate_one_patient (patient record as message body)
PATIENT_MESSAGE_TYPE = "PATIENT"


def build_one_batch(
    batch_size: int,
    patient_start_base: int,
    next_id: Synchronized,
    duplicate_ratio: float = DUPLICATE_RATIO,
) -> list[Record]:
    """Build one batch of records (same logic as run_producer). Call in order for each producer before starting threads so build order is deterministic."""
    batch: list[Record] = []
    for _ in range(batch_size):
        if random.random() < duplicate_ratio:
            with next_id.get_lock():
                existing_max = next_id.value - 1
            if existing_max < patient_start_base:
                with next_id.get_lock():
                    ordinal = next_id.value
                    next_id.value += 1
                is_original = True
            else:
                ordinal = patient_start_base + random.randint(
                    0, existing_max - patient_start_base
                )
                is_original = False
        else:
            with next_id.get_lock():
                ordinal = next_id.value
                next_id.value += 1
            is_original = True
        p = generate_one_patient(ordinal, is_original)
        record: Record = (
            p["PATIENT_ID"],
            PATIENT_MESSAGE_TYPE,
            json.dumps(p, default=str),
            p["is_original"],
        )
        batch.append(record)
    return batch


def run_batch_producer(
    insertion_queue: queue.Queue[Record | None],
    initial_batch: list[Record],
    batch_size: int,
    patient_start_base: int,
    next_id: Union[Synchronized, Any],
    recv_trigger: queue.Queue[None],
    send_trigger: queue.Queue[None],
    stop_event: Any,
    duplicate_ratio: float = DUPLICATE_RATIO,
) -> None:
    """Round-robin producer: wait for signal, put pre-built batch onto queue (only insertions between recv and send), signal next producer, then build next batch.
    Repeats until stop_event is set. Use build_one_batch in order before starting threads to pre-build initial batches."""
    batch = initial_batch
    while not stop_event.is_set():
        try:
            recv_trigger.get(timeout=0.1)
        except queue.Empty:
            continue
        if stop_event.is_set():
            break
        # Only insertions to the queue between recv and send (mimic Go).
        for record in batch:
            insertion_queue.put(record)
        send_trigger.put(None)
        # Build next batch for next iteration (outside the recv/send window).
        batch = build_one_batch(batch_size, patient_start_base, next_id, duplicate_ratio)


def run_producer(
    duration_sec: float,
    insertion_queue: queue.Queue[Record | None],
    num_workers: int,
    target_rps: int,
    patient_start_base: int,
    next_id: Synchronized,
    sentinels: int | None = None,
    max_records: int | None = None,
) -> None:
    """Enqueue single records at target_rps until duration_sec or max_records (if set), then enqueue sentinels.
    next_id is a shared multiprocessing.Value('q') holding the next patient ordinal (atomic counter).
    patient_start_base is the first ordinal in the range; duplicates are chosen in [patient_start_base, current_max].
    sentinels: number of INSERTION_SENTINEL to put at end; None = num_workers (single producer), 0 = none (multi-producer).
    max_records: if set, stop after this many records (rate-limited by target_rps); duration_sec is ignored when set."""
    if sentinels is None:
        sentinels = num_workers
    interval = 1.0 / target_rps if target_rps > 0 else 0.0
    start = time.perf_counter()
    next_put_at = start
    count = 0
    while True:
        if max_records is not None and count >= max_records:
            break
        if max_records is None and time.perf_counter() - start >= duration_sec:
            break
        now = time.perf_counter()
        if now >= next_put_at:
            if random.random() < DUPLICATE_RATIO:
                with next_id.get_lock():
                    existing_max = next_id.value - 1
                if existing_max < patient_start_base:
                    with next_id.get_lock():
                        ordinal = next_id.value
                        next_id.value += 1
                    is_original = True
                else:
                    ordinal = patient_start_base + random.randint(
                        0, existing_max - patient_start_base
                    )
                    is_original = False
            else:
                with next_id.get_lock():
                    ordinal = next_id.value
                    next_id.value += 1
                is_original = True
            p = generate_one_patient(ordinal, is_original)
            record: Record = (
                p["PATIENT_ID"],
                PATIENT_MESSAGE_TYPE,
                json.dumps(p, default=str),
                p["is_original"],
            )
            insertion_queue.put(record)
            count += 1
            next_put_at = next_put_at + interval
            if next_put_at < now:
                next_put_at = now + interval
        else:
            time.sleep(min(0.001, next_put_at - now))
    for _ in range(sentinels):
        insertion_queue.put(INSERTION_SENTINEL)

File: benchmark-go/internal/producer/producer.go
package producer

import (
	"context"
	"encoding/json"
	"math/rand"
	"sync/atomic"

	"github.com/db-benchmarking/internal/model"
	"github.com/db-benchmarking/internal/patientgen"
)

const patientMessageType = "PATIENT"

// BuildInsertPair builds one InsertPair (one batch split into originals and unique duplicates).
// Call this for each producer in order before starting producer goroutines so batch building order is deterministic.
func BuildInsertPair(batchSize int, patientStartBase int, nextID *atomic.Int64, duplicateRatio float64) *model.InsertPair {
	return buildInsertPair(batchSize, patientStartBase, nextID, duplicateRatio)
}

func buildInsertPair(batchSize int, patientStartBase int, nextID *atomic.Int64, duplicateRatio float64) *model.InsertPair {
	batch := make([]*model.Record, 0, batchSize)
	for len(batch) < batchSize {
		var ordinal int
		var isOriginal bool
		if rand.Float64() < duplicateRatio {
			existingMax := nextID.Load() - 1
			if existingMax < int64(patientStartBase) {
				ordinal = int(nextID.Add(1) - 1)
				isOriginal = true
			} else {
				ordinal = patientStartBase + rand.Intn(int(existingMax)-patientStartBase+1)
				isOriginal = false
			}
		} else {
			ordinal = int(nextID.Add(1) - 1)
			isOriginal = true
		}
		p := patientgen.GenerateOnePatient(ordinal, isOriginal)
		jsonMsg, _ := p.ToJSON()
		batch = append(batch, &model.Record{
			PatientID:   p.PatientID,
			MessageType: patientMessageType,
			JSONMessage: jsonMsg,
			IsOriginal:  p.IsOriginal,
		})
	}
	var originals []*model.Record
	for _, r := range batch {
		if r != nil && r.IsOriginal {
			originals = append(originals, r)
		}
	}
	seen := make(map[string]struct{})
	var duplicates []*model.Record
	for _, r := range batch {
		if r == nil || r.IsOriginal {
			continue
		}
		key := r.PatientID + "\x00" + r.MessageType + "\x00" + r.JSONMessage
		if _, ok := seen[key]; ok {
			continue
		}
		seen[key] = struct{}{}
		duplicates = append(duplicates, r)
	}
	return &model.InsertPair{Originals: originals, Duplicates: duplicates}
}

// Run produces batches of records and enqueues them until ctx is cancelled.
// initialPair is the pre-built pair for this producer (from BuildInsertPair, called in order before starting goroutines).
// The loop only does: recv -> insert to queue -> send -> build next pair. Between recvCh and sendCh only the queue insertion runs.
func Run(
	ctx context.Context,
	insertionQueue chan *model.InsertPair,
	initialPair *model.InsertPair,
	batchSize int,
	patientStartBase int,
	nextID *atomic.Int64,
	duplicateRatio float64,
	recvCh <-chan struct{},
	sendCh chan<- struct{},
) {
	if batchSize <= 0 {
		return
	}
	pair := initialPair

	for {
		select {
		case <-ctx.Done():
			select {
			case <-recvCh:
				sendCh <- struct{}{}
			default:
			}
			return
		case <-recvCh:
		}
		if ctx.Err() != nil {
			sendCh <- struct{}{}
			return
		}
		// Only insertion to the queue between recvCh and sendCh.
		insertionQueue <- pair
		sendCh <- struct{}{}
		// Build next pair for the next iteration (outside the recv/send window).
		pair = buildInsertPair(batchSize, patientStartBase, nextID, duplicateRatio)
	}
}

func init() {
	_, _ = json.Marshal(patientgen.PatientRecord{})
}
